R/java_unpack.R

```diff
@@ -0,0 +1,79 @@
+#' Unpack a Java distribution file into cache directory
+#' 
+#' @inheritParams java_install
+#' @inheritParams global_quiet_param
+#' @return A `character` vector containing of length 1 containing the path to the unpacked Java directory.
+#' @keywords internal
+java_unpack <- function(
+  java_distrib_path,
+  quiet = FALSE
+) {
+  platforms <- c("windows", "linux", "macos")
+  architectures <- c("x64", "aarch64", "arm64")
+  java_versions <- getOption("rJavaEnv.valid_major_java_versions")
+
+  # Extract information from the file name
+  filename <- basename(java_distrib_path)
+  parts <- strsplit(gsub("\\.tar\\.gz|\\.zip", "", filename), "-")[[1]]
+
+  # Guess the version, architecture, and platform
+  version <- parts[vapply(parts, function(x) x %in% java_versions, logical(1))][1]
+  arch <- parts[vapply(parts, function(x) x %in% architectures, logical(1))][1]
+  platform <- parts[vapply(parts, function(x) x %in% platforms, logical(1))][1]
+
+  if (is.na(version)) stop(cli::cli_abort("Unable to detect Java version from filename.", .envir = environment()))
+  if (is.na(arch)) stop(cli::cli_abort("Unable to detect architecture from filename.", .envir = environment()))
+  if (is.na(platform)) stop(cli::cli_abort("Unable to detect platform from filename.", .envir = environment()))
```

Comment:
```suggestion
  if (is.na(version)) cli::cli_abort("Unable to detect Java version from filename.")
  if (is.na(arch)) cli::cli_abort("Unable to detect architecture from filename.")
  if (is.na(platform)) cli::cli_abort("Unable to detect platform from filename.")
```

You only need to pass `envir` when you're using a helper that throws and error and want to make sure the error message references the function that the user actually called.